content/docs/guides/author_develop/pwa-amp/amp-in-pwa.md

```diff
@@ -0,0 +1,102 @@
+
+##  Include "Shadow AMP" in your Progressive Web App
+
+The first step is to include a special version of AMP we call “Shadow AMP” in your Progressive Web App. Yes, that’s right – you load the AMP library in the top level page, but it won’t actually control the top level content. It will only “amplify” the portions of our page that you tell it to.
+
+Include Shadow AMP in the head of your page, like so:
+
+[sourcecode:html]
+<!-- Asynchronously load the AMP-with-Shadow-DOM runtime library. -->
+<script async src="https://cdn.ampproject.org/shadow-v0.js"></script>
+[/sourcecode]
+
+## Handle navigation in your Progressive Web App
+
+You’ll still need to implement this step manually. After all, it's up to you how you present links to content in your navigation concept. A number of lists? A bunch of cards?
+
+In a common scenario, you’d fetch some JSON that returns ordered URLs with some metadata. In the end, you should end up with a function callback that fires when the user clicks on one of the links, and said callback should include the URL of the requested AMP page. If you have that, you’re all set for the final step.
+
+## Use the Shadow API to render a page inline
+
+Finally, when you want to display content after a user action, it's time to fetch the relevant AMP document and let Shadow AMP take over. First, implement a function to fetch the page, similar to this one:
+
+[sourcecode:javascript]
+function fetchDocument(url) {
+
+  // unfortunately fetch() does not support retrieving documents,
+  // so we have to resort to good old XMLHttpRequest.
+  var xhr = new XMLHttpRequest();
+
+  return new Promise(function(resolve, reject) {
+    xhr.open('GET', url, true);
+    xhr.responseType = 'document';
+    xhr.setRequestHeader('Accept', 'text/html');
+    xhr.onload = function() {
+      // .responseXML contains a ready-to-use Document object
+      resolve(xhr.responseXML);
+    };
+    xhr.send();
+  });
+}
+[/sourcecode]
+
+{% call callout('Important', type='caution') %}
+To simplify the above code example, we skipped over error handling. You should always make sure to catch and handle errors gracefully.
+{% endcall %}
+
+Now that we have our ready-to-use `Document` object, it's time to let AMP take over and render it. Get a reference to the DOM element that serves as container for the AMP document, then call `AMP.attachShadowDoc()`, like so:
+
+[sourcecode:javascript]
+// This can be any DOM element
+var container = document.getElementById('#container');
+
+// The AMP page you want to display
+var url = "https://my-domain/amp/an-article.html";
+
+// Use our fetchDocument method to get the doc
+fetchDocument(url).then(function(doc) {
+  // Let AMP take over and render the page
+  var ampedDoc = AMP.attachShadowDoc(container, doc, url);
```

Comment:
So, since we include the `shadow-v0.js` above as `async` (which we should always do), the problem here is that we don't know when `AMP` will be available. A simple answer is "when `shadow-v0.js` is downloaded and initialized". A practical version of this is:
```
window.AMP = window.AMP || []).push(function(AMP) {
  // AMP is now available.
});
```

This is obviously very visually displeasing, but unfortunately Web doesn't give us any other option. This `Array.push-as-Promise.then` pattern has become the leading way to do this.

Typically we'd wrap this call into a promise. Then, of course, promises are not available everywhere yet. AMP polyfills them, but it's not available until AMP is downloaded and initialized. Catch 22. But that's a topic for another discussion.